zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+                                                            recipient_id__in=stream_recipient_ids)
+    stream_subscription_dicts = [model_to_dict(x) for x in stream_subscription_query]
+
+    if include_private:
+        # First we get all huddles involving someone in the realm.
+        realm_huddle_subs = Subscription.objects.select_related("recipient").filter(recipient__type=3,
+                                                                                    user_profile__in=user_profile_ids)
+        realm_huddle_recipient_ids = set(sub.recipient_id for sub in realm_huddle_subs)
+
+        # Mark all Huddles whose recipient ID contains a cross-realm user.
+        unsafe_huddle_recipient_ids = set()
+        for sub in Subscription.objects.select_related().filter(recipient__in=realm_huddle_recipient_ids):
+            if sub.user_profile.realm != realm:
+                # In almost every case the other realm will be zulip.com
+                unsafe_huddle_recipient_ids.add(sub.recipient_id)
+
+        # Now filter down to just those huddles that are entirely within the realm.
+        huddle_subs = [sub for sub in realm_huddle_subs if sub.recipient_id not in unsafe_huddle_recipient_ids]
+        huddle_recipient_ids = set(sub.recipient_id for sub in huddle_subs)
+        huddle_ids = set(sub.recipient.type_id for sub in huddle_subs)
+
+        huddle_subscription_dicts = [model_to_dict(x) for x in huddle_subs]
+        huddle_recipients = [model_to_dict(r) for r in Recipient.objects.filter(id__in=huddle_recipient_ids)]
+        response['zerver_huddle'] = [model_to_dict(h) for h in Huddle.objects.filter(id__in=huddle_ids)]
+    else:
+        huddle_recipients = []
+        huddle_subscription_dicts = []
+
+    response["zerver_recipient"] = user_recipients + stream_recipients + huddle_recipients
+    response["zerver_subscription"] = user_subscription_dicts + stream_subscription_dicts + huddle_subscription_dicts
+
+    response["zerver_attachment"] = [model_to_dict(x) for x in
+                                     Attachment.objects.filter(realm=realm)]
+    floatify_datetime_fields(response, 'zerver_attachment', 'create_time')
+
+def fetch_usermessages(realm, message_ids, user_profile_ids, message_filename):
+    # UserMessage export security rule: You can export UserMessages
+    # for the messages you exported for the users in your realm.
+    #
+    # TODO: Optimize this query to do the same thing more efficiently
+    user_message_query = UserMessage.objects.filter(user_profile__realm=realm,
+                                                    message_id__in=message_ids)
+    user_message_chunk = [model_to_dict(um) for um in user_message_query if
+                          um.user_profile_id in user_profile_ids]
+    logging.info("Fetched UserMessages for %s" % (message_filename,))
+    return user_message_chunk
+
+def export_usermessages_batch(input_path, output_path):
```

Comment:
maybe add comment that this runs as a worker, which presumably explains why we pass in a JSON file instead of having the caller just pass in data to us